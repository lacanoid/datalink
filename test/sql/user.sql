\pset null _null_

SET client_min_messages = notice;
SET search_path = public,datalink;

begin;

create role datalink_test_user;
grant all on schema public to datalink_test_user;
set role datalink_test_user;

create table user_links (
  id serial,
  link datalink,
  link2 datalink
);

reset role;

update datalink.columns
   set link_control='FILE', integrity='ALL',
       read_access='DB', write_access='TOKEN',
       recovery='YES', on_unlink='RESTORE'
 where table_name='user_links' and column_name='link';

update datalink.columns
   set link_control='FILE', integrity='ALL'
 where table_name='user_links' and column_name='link2';

set role datalink_test_user;

insert into user_links (link)
values (dlvalue('/var/www/datalink/CHANGELOG.md','FS','Sample file datalink 1'));

update user_links
   set link = dlnewcopy(link);

select * 
  from datalink.linked_files;

update user_links set link = null;

truncate user_links;

reset role;
drop table user_links;
revoke all on schema public from datalink_test_user;
drop role datalink_test_user;

abort;
