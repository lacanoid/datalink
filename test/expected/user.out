\pset null _null_
SET client_min_messages = notice;
SET search_path = public,datalink;
begin;
create role datalink_test_user;
grant all on schema public to datalink_test_user;
set role datalink_test_user;
create table user_links (
  id serial,
  link datalink,
  link2 datalink
);
reset role;
update datalink.columns
   set link_control='FILE', integrity='ALL',
       read_access='DB', write_access='TOKEN',
       recovery='YES', on_unlink='RESTORE'
 where table_name='user_links' and column_name='link';
NOTICE:  DATALINK DDL:TRIGGER on user_links
update datalink.columns
   set link_control='FILE', integrity='ALL'
 where table_name='user_links' and column_name='link2';
set role datalink_test_user;
insert into user_links (link)
values (dlvalue('/var/www/datalink/CHANGELOG.md','FS','Sample file datalink 1'));
WARNING:  DATALINK WARNING - datalinker required
HINT:  make sure pg_datalinker process is running to finalize your commits
NOTICE:  DATALINK LINK:/var/www/datalink/CHANGELOG.md
update user_links
   set link = dlnewcopy(link);
NOTICE:  DATALINK UNLINK:/var/www/datalink/CHANGELOG.md
WARNING:  DATALINK WARNING - datalinker required
HINT:  make sure pg_datalinker process is running to finalize your commits
NOTICE:  DATALINK LINK:/var/www/datalink/CHANGELOG.md
select * 
  from datalink.linked_files;
              path              | state | read_access | write_access | recovery | on_unlink |  regclass  | attname |       owner        |  err   
--------------------------------+-------+-------------+--------------+----------+-----------+------------+---------+--------------------+--------
 /var/www/datalink/CHANGELOG.md | LINK  | DB          | TOKEN        | YES      | RESTORE   | user_links | link    | datalink_test_user | _null_
(1 row)

update user_links set link = null;
NOTICE:  DATALINK UNLINK:/var/www/datalink/CHANGELOG.md
truncate user_links;
savepoint p1;
-- test permissions on file functions
select * from datalink.read_text('/etc/passwd',1,4); -- should fail
ERROR:  DATALINK EXCEPTION - SELECT permission denied on directory.
FILE:  /etc/passwd
CONTEXT:  PL/Perl function "read_text"
rollback to p1;
select * from datalink.read_text('/var/www/datalink/CHANGELOG.md',1,12);
  read_text   
--------------
 Version 0.23
(1 row)

reset role;
drop table user_links;
revoke all on schema public from datalink_test_user;
drop role datalink_test_user;
abort;
